server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/SpanServiceElasticsearch.java

```diff
@@ -192,6 +201,83 @@ public void clear(String tenantId) {
         }
     }
 
+    @Override
+    public Trace getTraceFragment(String tenantId, String id) {
+        Span span = getSpan(tenantId, id);
+
+        Trace trace = spanToTrace(span);
+        LinkedList<Node> nodes = recursiveTraceFragment(tenantId, id);
+        nodes.push(spanToNode(span));
+        trace.setNodes(nodes);
+
+        return trace;
+    }
+
+
+    private LinkedList<Node> recursiveTraceFragment(String tenantId, String spanId) {
+        List<Span> spanChildren = getChildren(tenantId, spanId);
+        if (spanChildren == null) {
+            return new LinkedList<>();
+        }
+
+        LinkedList<Node> nodes = new LinkedList<>();
+
+        for (Span span: spanChildren) {
+            if (span.serverSpan()) {
+                continue;
+            }
+
+            InteractionNode node = spanToNode(span);
+            node.setNodes(recursiveTraceFragment(tenantId, span.getId()));
```

Comment:
A better way to stop the recursion may be to check that the span is not a clienSpan(), instead of the continue on a serverSpan().